solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -766,12 +766,16 @@ public Document doc(int i, Set<String> fields) throws IOException {
     }
 
     final DirectoryReader reader = getIndexReader();
-    if (!enableLazyFieldLoading || fields == null) {
-      d = reader.document(i);
+    if (fields != null) {
+      if (enableLazyFieldLoading) {
+        final SetNonLazyFieldSelector visitor = new SetNonLazyFieldSelector(fields, reader, i);
+        reader.document(i, visitor);
+        d = visitor.doc;
+      } else {
+        d = reader.document(i, fields);
```

Comment:
Right. I *think* I addressed that in the last commit, since only full documents are cached now. The problem is that the overhead of doing this might be unacceptable.